### PR TITLE
Implicit Arguments

### PR DESCRIPTION
**Pull request by CodaFi**
### What's in this pull request?

A few small fixes to the diagnostic engine and implicit argument support.

### CODE REVIEW

File: Sources/Moho/NameBinding.swift [hunk 1]
```diff
@@ -335,7 +335,7 @@ extension NameBinding {
   }
 
   private func checkNotReserved(_ n: Name) -> Bool {
-    if Set(["Type"]).contains(n.string) {
+    if Set([TokenKind.typeKeyword.text]).contains(n.string) {
```
Comment (harlanhaskins) [inline Sources/Moho/NameBinding.swift:338]: Just in case we change `Type` to `Set` 😉

File: Sources/Moho/ScopeCheck.swift [hunk 2]
```diff
@@ -269,15 +286,30 @@ extension NameBinding {
...
+                                                   DeclaredPattern.wild,
+                                                   self.exprToDeclPattern)
+    if let err = maybeError {
+      switch err {
+      case let .extraneousArguments(expected, have, implicit, leftoverStart):
+        let diag: Diagnostic.Message =
+          .tooManyPatternsInLHS(expected, have, implicit)
+        self.engine.diagnose(diag, node: syntax) {
+          for i in leftoverStart..<syntax.count {
+            $0.note(.ignoringExcessPattern, node: syntax[i])
```
Comment (harlanhaskins) [inline Sources/Moho/ScopeCheck.swift:308]: I would go ahead and reprint the pattern's source text here...

File: Sources/Moho/ScopeCheck.swift [hunk 3]
```diff
@@ -443,33 +491,77 @@ extension NameBinding {
...
+  }
+
+  private func computePlicity(_ syntax: ExprSyntax) -> [ArgumentPlicity] {
+    var plicities = [ArgumentPlicity]()
+    func go(_ syntax: ExprSyntax) {
+      switch syntax {
+      case let syntax as ReparsedApplicationExprSyntax:
+        let headExpr = syntax.head
+        let n = QualifiedName(ast: headExpr.name)
+        guard n.string == "_->_" else {
```
Comment (harlanhaskins) [inline Sources/Moho/ScopeCheck.swift]: Add a FIXME for this to Do This Properly™

**Review by harlanhaskins — APPROVED** [review]

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**